lib/push_notifications.rb

```diff
@@ -0,0 +1,51 @@
+class PushNotifications
+  KEY_PREFIX = "push-notification-subscription:".freeze
+  GCM_ENDPOINT = 'https://android.googleapis.com/gcm/send'.freeze
+
+  def self.registration_ids(user)
+    $redis.smembers(redis_key(user))
+  end
+
+  def self.subscribe(user, endpoint)
+    $redis.sadd(redis_key(user), extract_registration_id(endpoint))
+  end
+
+  def self.unsubscribe(user, endpoint)
+    $redis.srem(redis_key(user), extract_registration_id(endpoint))
+  end
+
+  def self.push(user)
+    self.registration_ids(user).each do |registration_id|
+      if !SiteSetting.gcm_api_key.blank?
+        headers = {
+          'Content-Type' => 'application/x-www-form-urlencoded;charset=UTF-8',
+          'Authorization' => "key=#{SiteSetting.gcm_api_key}"
+        }
+
+        body = URI.encode_www_form(
+          registration_id: registration_id,
+          collapse_key: Discourse.base_url
+        )
+
+        result = Excon.post(GCM_ENDPOINT, headers: headers, body: body)
+
+        if result.status == 200 && result.body.start_with?("Error")
+          Rails.logger.error("There was an error from Google Cloud Message: #{result.body}")
+        elsif result.status != 200
```

Comment:
TODO: Check the response status and unsubscribe registration ids that are no longer in use.